File: NinePSharp/messages/Twalk.cs
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

namespace NinePSharp.Messages;

// size[4] Twalk tag[2] fid[4] newfid[4] nwname[2] nwname*(wname[s])
public readonly struct Twalk : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Twalk;
    public ushort Tag { get; }

    public uint Fid { get; }
    public uint NewFid { get; }
    public string[] Wname { get; }

    public Twalk(ushort tag, uint fid, uint newFid, string[] wname)
    {
        Tag = tag;
        Fid = fid;
        NewFid = newFid;
        Wname = wname;
        // Size will be calculated during write if needed, 
        // but let's set a minimal valid size for now
        Size = (uint)(NinePConstants.HeaderSize + 4 + 4 + 2 + (wname?.Sum(s => 2 + s.Length) ?? 0));
    }

    public Twalk(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));

        int offset = NinePConstants.HeaderSize;

        Fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4));
        offset += 4;

        NewFid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4));
        offset += 4;

        ushort nwname = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(offset, 2));
        offset += 2;

        Wname = new string[nwname];
        for (int i = 0; i < nwname; i++)
        {
            Wname[i] = data.ReadString(ref offset);
        }
    }

    public void WriteTo(Span<byte> data)
    {
        data.WriteHeaders(Size, Tag, Type);
        int offset = NinePConstants.HeaderSize;

        BinaryPrimitives.WriteUInt32LittleEndian(data.Slice(offset, 4), Fid);
        offset += 4;

        BinaryPrimitives.WriteUInt32LittleEndian(data.Slice(offset, 4), NewFid);
        offset += 4;

        ushort nwname = (ushort)(Wname?.Length ?? 0);
        BinaryPrimitives.WriteUInt16LittleEndian(data.Slice(offset, 2), nwname);
        offset += 2;

        for (int i = 0; i < nwname; i++)
        {
            data.WriteString(Wname![i], ref offset);
        }
    }
}

File: NinePSharp.Messages.FSharp/messages/Twalk.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Twalk =
    { Size: uint32
      Tag: uint16
      Fid: uint32
      NewFid: uint32
      Wname: string[] }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Twalk
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Twalk span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Fid); offset <- offset + 4
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.NewFid); offset <- offset + 4
            let len = match this.Wname with | null -> 0us | x -> uint16 x.Length
            BinaryPrimitives.WriteUInt16LittleEndian(span.Slice(offset, 2), len); offset <- offset + 2
            if not (isNull this.Wname) then
                for name in this.Wname do
                    writeString name span &offset

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let newfid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let nwname = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(offset, 2))
        offset <- offset + 2
        let wname = Array.zeroCreate (int nwname)
        for i = 0 to int nwname - 1 do
            wname.[i] <- readString data &offset
        { Size = size; Tag = tag; Fid = fid; NewFid = newfid; Wname = wname }

